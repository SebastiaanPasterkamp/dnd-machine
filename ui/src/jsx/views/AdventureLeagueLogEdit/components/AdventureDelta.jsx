import React from 'react';
import PropTypes from 'prop-types';

import CalculatorInputField from '../../../components/CalculatorInputField.jsx';
import ControlGroup from '../../../components/ControlGroup.jsx';
import InputField from '../../../components/InputField.jsx';
import Panel from '../../../components/Panel.jsx';

export class AdventureDelta extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.state = {};
    }

    onChange = (earned, formula) => {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }

        const total = starting + earned;
        this.setState(
            { formula, total },
            () => setState({ starting, earned, total })
        );
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;
        const {
            formula = earned,
        } = this.state;

        return (
            <Panel
                header={label}
                className={className}
            >
                <ControlGroup label="Starting">
                    <InputField
                        placeholder="Starting..."
                        disabled={true}
                        type="number"
                        value={starting}
                    />
                </ControlGroup>
                <ControlGroup label="Earned">
                    <CalculatorInputField
                        placeholder="Earned..."
                        data-field="earned"
                        value={earned}
                        disabled={disabled}
                        setState={this.onChange}
                    />
                </ControlGroup>
                <ControlGroup label="Total">
                    <InputField
                        placeholder="Total..."
                        disabled={true}
                        type="number"
                        value={earned ? total : starting}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureDelta.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.number,
    earned: PropTypes.number,
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureDelta.defaultProps = {
    starting: 0,
    earned: 0,
    className: '',
    disabled: false,
};

export default AdventureDelta;
