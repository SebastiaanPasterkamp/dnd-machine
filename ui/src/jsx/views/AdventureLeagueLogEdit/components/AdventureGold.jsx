import React from 'react';
import PropTypes from 'prop-types';
import { mergeWith } from 'lodash/fp';

import Coinage from '../../../components/Coinage.jsx';
import ControlGroup from '../../../components/ControlGroup.jsx';
import CostEditor from '../../../components/CostEditor.jsx';
import Panel from '../../../components/Panel.jsx';

export class AdventureGold extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.state = {};
    }

    onChange = (earned) => {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }

        const total = mergeWith(
            (dst, src) => ((dst || 0) + src),
            starting,
            earned,
        );
        this.setState(
            { total },
            () => setState({ starting, earned, total })
        );
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;

        return (
            <Panel
                className={className}
                header={label}
            >
                <ControlGroup label="Starting">
                    <Coinage
                        {...starting}
                        className="nice-form-control"
                        extended="1"
                    />
                </ControlGroup>

                <ControlGroup label="Earned">
                    <CostEditor
                        value={earned}
                        disabled={disabled}
                        setState={this.onChange}
                    />
                </ControlGroup>

                <ControlGroup label="Total">
                    <Coinage
                        {...total}
                        className="nice-form-control"
                        extended="1"
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureGold.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.objectOf(PropTypes.number),
    earned: PropTypes.objectOf(PropTypes.number),
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureGold.defaultProps = {
    starting: {},
    earned: {},
    className: '',
    disabled: false,
};

export default AdventureGold;
