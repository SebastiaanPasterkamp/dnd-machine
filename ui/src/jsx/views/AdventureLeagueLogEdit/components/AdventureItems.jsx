import React from 'react';
import PropTypes from 'prop-types';
import {
    filter,
    isObject,
    map,
} from 'lodash/fp';

import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import Panel from '../../../components/Panel';
import ListComponent from '../../../components/ListComponent';

export class AdventureItems extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.initalItem = {
            value: '',
        };
        this.onChange = this.onChange.bind(this);
    }

    onChange(items) {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }
        const earned = filter(item => item)(
            map(
                item => ( isObject(item) ? item.value : item )
            )(items)
        );
        const total = starting + earned.length;
        setState({ starting, earned, total });
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;

        return (
            <Panel
                className={className}
                header={label}
            >
                <ControlGroup label="Starting">
                    <InputField
                        placeholder="Starting..."
                        disabled={true}
                        type="number"
                        value={starting}
                    />
                </ControlGroup>

                <ControlGroup label="Obtained">
                    <ListComponent
                        newItem='auto'
                        list={map( value => ({ value }) )(earned)}
                        component={InputField}
                        initialItem={this.initalItem}
                        disabled={disabled}
                        setState={!disabled
                            ? this.onChange
                            : null
                        }
                        componentProps={{ disabled }}
                    />
                </ControlGroup>

                <ControlGroup label="Total">
                    <InputField
                        placeholder="Total..."
                        disabled={true}
                        type="number"
                        value={total}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureItems.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.number,
    earned: PropTypes.arrayOf( PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
            item: PropTypes.string,
        }),
    ])),
    total: PropTypes.number,
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureItems.defaultProps = {
    starting: 0,
    earned: [],
    total: 0,
    className: '',
    disabled: false,
};

export default AdventureItems;
