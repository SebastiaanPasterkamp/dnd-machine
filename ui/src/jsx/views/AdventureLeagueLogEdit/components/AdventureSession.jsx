import React from 'react';
import PropTypes from 'prop-types';

import { memoize } from '../../../utils';

import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import Panel from '../../../components/Panel';

export class AdventureSession extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
    }

    onChange(field) {
        return this.memoize(
            field,
            (value) => this.props.setState({ [field]: value })
        );
    }

    render() {
        const {
            name, id, date, dm_name, dm_dci, setState,
        } = this.props;

        return (
            <Panel
                className="al-log-edit__session"
                header="Adventure"
            >
                <ControlGroup label="Name">
                    <InputField
                        data-field="name"
                        placeholder="Name..."
                        value={name}
                        setState={this.onChange('name')}
                    />
                </ControlGroup>

                <ControlGroup label="Session #">
                    <InputField
                        data-field="id"
                        placeholder="Session #..."
                        value={id}
                        setState={this.onChange('id')}
                    />
                </ControlGroup>

                <ControlGroup label="Date">
                    <InputField
                        data-field="date"
                        placeholder="Date..."
                        type="date"
                        value={date}
                        setState={this.onChange('date')}
                    />
                </ControlGroup>

                <ControlGroup label="DM Name">
                    <InputField
                        data-field="dm_name"
                        placeholder="DM Name..."
                        value={dm_name}
                        setState={this.onChange('dm_name')}
                    />
                </ControlGroup>

                <ControlGroup label="DM DCI">
                    <InputField
                        data-field="dm_dci"
                        placeholder="DM DCI..."
                        value={dm_dci}
                        setState={this.onChange('dm_dci')}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureSession.propTypes = {
    setState: PropTypes.func,
    name: PropTypes.string,
    id: PropTypes.string,
    date: PropTypes.string,
    dm_name: PropTypes.string,
    dm_dci: PropTypes.string,
};

AdventureSession.defaultProps = {
    name: '',
    id: '',
    date: '',
    dm_name: '',
    dm_dci: '',
};

export default AdventureSession;
