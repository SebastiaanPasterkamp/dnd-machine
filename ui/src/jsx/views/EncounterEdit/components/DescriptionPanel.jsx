import React from 'react';
import PropTypes from 'prop-types';
import {
    values,
} from 'lodash/fp';

import { memoize } from '../../../utils';

import ListDataWrapper from '../../../hocs/ListDataWrapper';
import ObjectDataListWrapper from '../../../hocs/ObjectDataListWrapper';

import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import Panel from '../../../components/Panel';
import SingleSelect from '../../../components/SingleSelect';
import MarkdownTextField from '../../../components/MarkdownTextField';

export class DescriptionPanel extends React.Component
{
    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
        this.onFieldChange = this.onFieldChange.bind(this);
    }

    onFieldChange(field) {
        const { setState } = this.props;
        return this.memoize(
            field,
            (value) => setState({ [field]: value })
        );
    }

    render() {
        const { campaign_id, campaigns, currentCampaign, name, description } = this.props;

        return (
            <Panel
                key="description"
                className="encounter-edit__description"
                header="Description"
            >
                <ControlGroup label="Campaign">
                    <SingleSelect
                        selected={campaign_id}
                        defaultValue={currentCampaign ? currentCampaign.id : null}
                        items={values(campaigns)}
                        renderEmpty="Shared"
                        setState={this.onFieldChange('campaign_id')}
                    />
                </ControlGroup>
                <ControlGroup label="Name">
                    <InputField
                        placeholder="Name..."
                        value={name}
                        setState={this.onFieldChange('name')}
                    />
                </ControlGroup>
                <ControlGroup label="Description">
                    <MarkdownTextField
                        placeholder="Description..."
                        value={description}
                        rows={5}
                        setState={this.onFieldChange('description')}
                    />
                </ControlGroup>
            </Panel>
        );
    }
}

DescriptionPanel.propTypes = {
    setState: PropTypes.func.isRequired,
    campaign_id: PropTypes.number,
    campaigns: PropTypes.object,
    currentCampaign: PropTypes.object,
    name: PropTypes.string,
    description: PropTypes.string,
};

DescriptionPanel.defaultProps = {
    campaign_id: null,
    campaigns: {},
    currentCampaign: {},
    name: '',
    description: '',
};

export default ListDataWrapper(
    ObjectDataListWrapper(
        DescriptionPanel,
        {campaigns: {type: 'campaign'}}
    ),
    ['current_campaign'],
    null,
    { current_campaign: 'currentCampaign' }
);
