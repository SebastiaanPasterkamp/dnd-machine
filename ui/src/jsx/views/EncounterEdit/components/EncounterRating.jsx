import React from 'react';
import PropTypes from 'prop-types';

import utils from '../../../utils';

import XpRating from '../../../components/XpRating';

export const EncounterRating = ({ encounter, party }) => {
    if (!encounter || !party) {
        return null;
    }

    const {
        easy = 0,
        medium = 0,
        hard = 0,
        deadly = 0,
    } = party;

    const rating = utils.closest({
        irrelevant: 0,
        easy, medium, hard, deadly,
        insane: deadly + deadly - hard,
    }, encounter, 'irrelevant');

    if (rating === 'easy') {
        return (
            <span>
                Easy
                (<XpRating xpRating={easy} />)
            </span>
        );
    }
    if (rating === 'medium') {
        return (
            <span>
                Medium
                (<XpRating xpRating={medium} />)
            </span>
        );
    }
    if (rating === 'hard') {
        return (
            <span>
                Hard
                (<XpRating xpRating={hard} />)
            </span>
        );
    }
    if (rating === 'deadly') {
        return (
            <span>
                Deadly
                (<XpRating xpRating={deadly} />)
            </span>
        );
    }
    if (rating === 'insane') {
        return (
            <span>
                Insane
                (&gt; <XpRating xpRating={deadly} />)
            </span>
        );
    }
    return (
        <span>
            Irrelevant
            (&lt; <XpRating xpRating={easy} />)
        </span>
    );
};

EncounterRating.propTypes = {
    encounter: PropTypes.number,
    party: PropTypes.objectOf( PropTypes.number ),
};

EncounterRating.defaultProps = {
    encounter: 0,
    party: null,
};

export default EncounterRating;
