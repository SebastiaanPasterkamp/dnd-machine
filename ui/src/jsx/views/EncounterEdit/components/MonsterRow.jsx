import React from 'react';
import PropTypes from 'prop-types';

import { BaseLinkButton } from '../../../components/BaseLinkGroup/index';
import Bonus from '../../../components/Bonus';
import ChallengeRating from '../../../components/ChallengeRating';
import CountPrefix from '../../../components/CountPrefix';
import MonsterLabel from '../../../components/MonsterLabel';
import MonsterLinks from '../../../components/MonsterLinks';
import XpRating from '../../../components/XpRating';

const monsterLinks = [ 'view' ];

export const MonsterRow = function({
    count,
    monster: {
        id,
        name,
        average_damage,
        attack_bonus,
        spell_save_dc,
        armor_class,
        hit_points,
        challenge_rating_precise,
        xp_rating,
    },
    onAddMonsterButton,
    onRemoveMonsterButton,
}) {
    if (id === undefined) {
        return null;
    }

    return (
        <tr>
            <th>
                <CountPrefix count={count} alwaysCount={true}>
                    {name}
                </CountPrefix>

                <MonsterLinks altStyle={true} id={id} include={monsterLinks}>
                    <BaseLinkButton
                        icon="minus"
                        altStyle={true}
                        className="warning"
                        action={onRemoveMonsterButton(id)}
                    />
                    <BaseLinkButton
                        icon="plus"
                        altStyle={true}
                        className="good"
                        action={onAddMonsterButton(id)}
                    />
                </MonsterLinks>
            </th>
            <td>
                <ChallengeRating
                    challengeRating={challenge_rating_precise}
                    precise={true}
                />
                {count > 1 ? (
                    <span>(
                        <ChallengeRating
                            challengeRating={challenge_rating_precise * count}
                            precise={true}
                        />
                    )</span>
                ) : null}
            </td>
            <td>
                <XpRating
                    xpRating={xp_rating}
                />
                {count > 1 ? (
                    <span>(
                        <XpRating xpRating={xp_rating * count} />
                    )</span>
                ) : null}
            </td>
            <td>{hit_points}</td>
            <td>{armor_class}</td>
            <td>
                {average_damage}
                {count > 1 ? (
                    <span>({average_damage * count})</span>
                ) : null}
            </td>
            <td>
                {attack_bonus ? (
                    <span>
                        Hit &nbsp; <Bonus bonus={attack_bonus} />
                    </span>
                ) : null}
                {spell_save_dc ? (
                    <span>DC &ge; {spell_save_dc}</span>
                ) : null}
            </td>
        </tr>
    );
};

MonsterRow.propTypes = {
    count: PropTypes.number.isRequired,
    monster: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        average_damage: PropTypes.number,
        attack_bonus: PropTypes.number,
        spell_save_dc: PropTypes.number,
        armor_class: PropTypes.number,
        hit_points: PropTypes.number,
        challenge_rating_precise: PropTypes.number,
        xp_rating: PropTypes.number,
    }),
    onAddMonsterButton: PropTypes.func.isRequired,
    onRemoveMonsterButton: PropTypes.func.isRequired,
};

MonsterRow.defaultProps = {
    count: 0,
    monster: {},
};

export default MonsterRow;
