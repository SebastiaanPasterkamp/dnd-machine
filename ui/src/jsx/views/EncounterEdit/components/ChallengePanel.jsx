import React from 'react';
import PropTypes from 'prop-types';

import ChallengeRating from '../../../components/ChallengeRating';
import Panel from '../../../components/Panel';
import XpRating from '../../../components/XpRating';

import EncounterRating from './EncounterRating';

export const ChallengePanel = function( {
    modifier, hosted_party,
    challenge_rating, challenge_modified, challenge_rating_precise,
    xp, xp_modified, xp_rating,
} ) {
    return (
        <Panel
            key="challenge"
            className="encounter-edit__challenge"
            header="Challenge Rating Overview"
        >
            <tbody>
                <tr>
                    <th>Monster Modifier</th>
                    <th>Party Modifier</th>
                    <th>Final Modifier</th>
                </tr>
                <tr>
                    <td>{modifier.monster}</td>
                    <td>{modifier.party}</td>
                    <td>{modifier.total}</td>
                </tr>
                <tr>
                    <th>Challenge Rating</th>
                    <th>Modified for party</th>
                    <th>Exact Challenge Rating</th>
                </tr>
                <tr>
                    <td>
                        <ChallengeRating
                            challengeRating={challenge_rating}
                        />
                        &nbsp;/&nbsp;
                        <XpRating
                            xpRating={xp}
                        />
                    </td>
                    {hosted_party && hosted_party.size ? (
                        <td>
                            <ChallengeRating
                                challengeRating={challenge_modified}
                            />
                            &nbsp;/&nbsp;
                            <XpRating
                                xpRating={xp_modified}
                            />
                        </td>
                    ) : (
                        <td>&mdash;</td>
                    )}
                    <td>
                        <ChallengeRating
                            challengeRating={challenge_rating_precise}
                            precise={true}
                        />
                        &nbsp;/&nbsp;
                        <XpRating
                            xpRating={xp_rating}
                            precise={true}
                        />
                    </td>
                </tr>
                {hosted_party && hosted_party.size ? (
                    <React.Fragment>
                        <tr>
                            <th>Party Size</th>
                            <th>XP / Party Member</th>
                            <th>Challenge Classification</th>
                        </tr>
                        <tr>
                            <td>{hosted_party.size}</td>
                            <td>
                                <XpRating
                                    xpRating={xp / hosted_party.size}
                                />
                            </td>
                            <td>
                                <EncounterRating
                                    encounter={xp_rating}
                                    party={hosted_party.challenge}
                                />
                            </td>
                        </tr>
                    </React.Fragment>
                ) : null}
            </tbody>
        </Panel>
    );
}

ChallengePanel.propTypes = {
    modifier: PropTypes.shape({
        monster: PropTypes.number.isRequired,
        party: PropTypes.number.isRequired,
        total: PropTypes.number.isRequired,
    }),
    challenge_rating: PropTypes.number,
    challenge_modified: PropTypes.number,
    challenge_rating_precise: PropTypes.number,
    xp: PropTypes.number,
    xp_modified: PropTypes.number,
    xp_rating: PropTypes.number,
    hosted_party: PropTypes.shape({
        id: PropTypes.number,
        size: PropTypes.number.isRequired,
        challenge: PropTypes.shape({
            easy: PropTypes.number.isRequired,
            medium: PropTypes.number.isRequired,
            hard: PropTypes.number.isRequired,
            deadly: PropTypes.number.isRequired,
        }),
    }),
};

ChallengePanel.defaultProps = {
    modifier: {
        monster: 1.0,
        party: 1.0,
        total: 1.0,
    },
    challenge_rating: 0,
    challenge_modified: 0,
    challenge_rating_precise: 0,
    xp: 0,
    xp_modified: 0,
    xp_rating: 0,
    hosted_party: null,
};

export default ChallengePanel;
