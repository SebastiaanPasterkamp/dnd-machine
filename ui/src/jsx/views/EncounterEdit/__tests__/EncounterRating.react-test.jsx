import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-enzyme';
import renderer from 'react-test-renderer';

import { EncounterRating } from '../components/EncounterRating';

describe('View EncounterRating', () => {
    const easyProps = {
        encounter: 6,
        party: {
            easy: 10,
            medium: 20,
            hard: 30,
            deadly: 40,
        },
    };
    const mediumProps = {
        encounter: 62,
        party: {
            easy: 25,
            medium: 50,
            hard: 75,
            deadly: 100,
        },
    };
    const insaneProps = {
        encounter: 230,
        party: {
            easy: 50,
            medium: 100,
            hard: 150,
            deadly: 200,
        },
    };

    it('should render w/ minimum props', () => {
        const tree = renderer.create(
            <EncounterRating
            />
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });

    describe('should render w/ various props;', () => {
        it('should identify an easy encounter', () => {
            const tree = renderer.create(
                <EncounterRating
                    {...easyProps}
                />
            );

            expect(tree.toJSON()).toMatchSnapshot();
        });
        it('should identify an medium encounter', () => {
            const tree = renderer.create(
                <EncounterRating
                    {...mediumProps}
                />
            );

            expect(tree.toJSON()).toMatchSnapshot();
        });
        it('should identify an insane encounter', () => {
            const tree = renderer.create(
                <EncounterRating
                    {...insaneProps}
                />
            );

            expect(tree.toJSON()).toMatchSnapshot();
        });
    });
});
