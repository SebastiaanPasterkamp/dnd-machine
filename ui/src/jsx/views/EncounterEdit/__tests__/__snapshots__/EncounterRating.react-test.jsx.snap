// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View EncounterRating should render w/ minimum props 1`] = `null`;

exports[`View EncounterRating should render w/ various props; should identify an easy encounter 1`] = `
<span>
  Easy (
  <div
    className="xp-rating inline"
  >
    10
     XP
  </div>
  )
</span>
`;

exports[`View EncounterRating should render w/ various props; should identify an insane encounter 1`] = `
<span>
  Insane (&gt; 
  <div
    className="xp-rating inline"
  >
    200
     XP
  </div>
  )
</span>
`;

exports[`View EncounterRating should render w/ various props; should identify an medium encounter 1`] = `
<span>
  Medium (
  <div
    className="xp-rating inline"
  >
    50
     XP
  </div>
  )
</span>
`;
